### ISSUE TITLE
Collection Manipulation Methods: sort, filter, diff, and other array function wrappers

### ISSUE THREAD

**Issue by jmauerhan** [ramsey/collection#13]
I propose adding new methods to `AbstractCollection` to allow easy manipulation of the contents. We could then use these methods in place of using for/foreach loops and temporary variables for building new collections of data. 

### Current Problem

For example, given a class Foo and a typed Collection FooCollection:
```php
Class Foo
{
    public $id;
    public $name;
    public $createdAt;

    public function __construct(int $id, string $name, DateTime $createdAt=null)
    {
        $this->id        = $id;
        $this->name      = $name;
        $this->createdAt = $createdAt;
    }
}

Class FooCollection extends \Ramsey\Collection\AbstractCollection
{
    public function getType()
    {
        return Foo::class;
    }
}
```

And you have created a collection of those Foos:
```php
$foos = [
    new Foo(1, 'bar', new DateTime('now')),
    new Foo(2, 'baz', new DateTime('1 week ago')),
    new Foo(3, 'fizz', new DateTime('3 months ago')),
    new Foo(4, 'buzz', new DateTime('2 days ago')),
];

$fooCollection = new FooCollection($foos);
```

And you have a method which was passed as an argument a `FooCollection`, and you want to sort your collection by the createdAt time, you'd have to do something like this:
```php
Class Bar{
    public function doSomething(FooCollection $fooCollection){
        $values = $fooCollection->toArray();
        usort($values, function($a, $b){
            return ($a->createdAt > $b->createdAt);
        });

        $fooCollection = new FooCollection($values);
        //Do something else with the Foos.
    }
}
```
### Proposed Solution

I propose a sort method would be much simpler for users:
```php
Class Bar{
    public function doSomething(FooCollection $fooCollection){
        $fooCollection->sort('createdAt');
        //Do something else with the Foos.
    }
}
```

# Suggested Methods

### sort
```php
/**
 * Orders the items by a property or return value of the method in the class
 * @param string property
 * @param string order 'desc'/'asc'
 * @return self
 */
sort($property, $order='asc')
```

#### Examples:
```php
$fooCollection->sort('createdAt', 'desc'); //sorts by property
$fooCollection->sort('getName'); //will call the method 
```

### filter
```php
/**
 * Filters out items using a callback method
 * @param function $callback
 * @return self
 */
filter($callback)
```

#### Examples:
```php
$foos = [
    new Foo(1, 'bar'),
    new Foo(2, 'baz'),
    new Foo(3, 'fizz'),
    new Foo(4, 'buzz'),
];
$fooCollection = new FooCollection($foos);

//$fooCollection will now contain only the 'bar' and 'baz' foos.
$fooCollection->filter(function($foo){
    return strlen($foo->name) <= 3;
});
```

### where
```php
/**
 * Filters to items where the property or return value of a method equals the given value
 * @param string property 
 * @param string value
 * @return self
 */
where($property, $value)
```

#### Examples:
```php
$fooCollection->where('name', 'foo'); //accesses a property
$fooCollection->where('getName', 'foo'); //will call the method 
```

### each
```php
/**
 * Applies a callback method to each item
 * @param function $callback
 * @return self
 */
each($callback)
```

#### Examples:
```php
//Every Foo's name in the collection will be updated to end in '.beta'
$fooCollection->each(function($foo){
    $foo->name .= '.beta';
});
```

### diff
```php
/**
 * Calculates the difference of two collections by comparing the objects - must be the same type
 * @param AbstractCollection
 */
diff($collection)
```

#### Examples:
```php
$someFoos = new FooCollection([
    new Foo(1, 'bar'),
    new Foo(2, 'baz')
]);

$moreFoos = new FooCollection([
    new Foo(2, 'baz'),
    new Foo(3, 'fizz'),
    new Foo(4, 'buzz'),
]);

// $diffCollection should contain the 'bar', 'fizz' and 'buzz' Foos - the difference between the two. 
$diffCollection = $someFoos->diff($moreFoos);
```

### intersect
```php
/**
 * Calculates the intersection of two collections by comparing the objects - must be the same type
 * @param AbstractCollection
 */
intersect($collection)
```

#### Examples:
```php
$someFoos = new FooCollection([
    new Foo(1, 'bar'),
    new Foo(2, 'baz')
]);

$moreFoos = new FooCollection([
    new Foo(2, 'baz'),
    new Foo(3, 'fizz'),
    new Foo(4, 'buzz'),
]);

// $intersectCollection should contain the 'baz' Foo - the intersection between the two. 
$intersectCollection = $someFoos->intersect($moreFoos);
```

### unique
```php
/**
 * Filters out non-unique values either by comparing the objects, or an optional property/method accessor
 * @param string property
 */
unique($property=null)
```

#### Examples:
```php
$foos = new FooCollection([
    new Foo(1, 'bar'),
    new Foo(2, 'baz')  //Original "baz"
    new Foo(3, 'baz'), //a duplicate "baz" with different id
    new Foo(4, 'fizz'),
    new Foo(5, 'buzz'), //Original "buzz"
    new Foo(5, 'buzz'), //Exact copy of "buzz" 
]);

// $uniqueFoos will contain 5 foos, the extra "buzz" having been removed.
$uniqueFoos = $uniqueFoos->unique();


// $uniqueFoos will contain 4 foos, the extra "buzz" and "baz" (#3) having been removed
$uniqueFoos = $uniqueFoos->unique('name');
```

### merge
```php
/**
 * Merges the two collections - must be the same type
 * @param AbstractCollection
 */
merge($collection)
```

#### Examples:
```php
$someFoos = new FooCollection([
    new Foo(1, 'bar'),
    new Foo(2, 'baz')
]);

$moreFoos = new FooCollection([
    new Foo(3, 'fizz'),
    new Foo(4, 'buzz'),
]);

// $allFoos should contain all 4 foos
$allFoos = $someFoos->intersect($moreFoos);
```

### merge
```php
/**
 * Merges the two collections - must be the same type
 * @param AbstractCollection
 */
merge($collection)
```

#### Examples:
```php
$someFoos = new FooCollection([
    new Foo(1, 'bar'),
    new Foo(2, 'baz')
]);

$moreFoos = new FooCollection([
    new Foo(3, 'fizz'),
    new Foo(4, 'buzz'),
]);

// $allFoos should contain all 4 foos
$allFoos = $someFoos->intersect($moreFoos);
```

### column
```php
/**
 * Wrapper for array_column: returns the values from a single property/method
 * @param string property 
 * @return array
 */
column($property)
```

#### Examples:
```php
$names = $fooCollection->column('name'); //accesses a property
$names = $fooCollection->column('getName'); //will call the method 
```

### first
```php
/**
 * Returns the first item in the collection
 * @return stdClass
 */
first()
```

#### Examples:
```php
$firstFoo = $fooCollection->first();
```

**Comment by ramsey** [comment]
I like all the suggested methods. I have a few notes and questions...

### Immutable objects?

Should the array objects be immutable? Should these methods alter the data on the object or return a brand new object according to the operation called?

I'm leaning toward changing the library so that the objects are all immutable and methods that change state return a new object, but I'm not sure if that's a good idea for everyone.

### sort

Should we use a third parameter for sort flags? Natural sort, locale string sort, numeric sort, etc.

### each

I think this is analogous to `array_map()`, so we'll have to decide whether to use `each()` or `map()` for the name. I could see arguments for either name, and I'll take a look at other languages to see what names they use.

### unique

When not passing a property/method we have to decide what constitutes an "exact copy." In your example, it looks like any object that contains all the same data, so identity doesn't matter. We may want to provide a stricter option where identity does matter, or we may want to default to checking identity by default and providing an option to relax that.

**Comment by jmauerhan** [comment]
I'm in favor of the immutable Collections. 

I also definitely think adding the flags for sort and comparison types would be useful. 

For unique, a flag to compare `==` vs `===` might help. I'd default to `===`. 

```
$foo  = new Foo(5, 'buzz');
$foo2 = new Foo(5, 'buzz');

var_dump($foo === $foo2); //false
var_dump($foo == $foo2); //true

$foo3 = $foo;
var_dump($foo === $foo3); //true
var_dump($foo == $foo3); //true
```

BTW, I told Ben this already, but my thought with this Issue was once we agree on the general approach, we can break each method out into it's own Issue so people can tackle them in small chunks for #hacktoberfest.

**Comment by shadowhand** [comment]
I would suggest not to go too overboard with this, because we already have [knapsack](https://packagist.org/packages/dusank/knapsack) that implements a number of highly performant and useful collection functions.

**Comment by jmauerhan** [comment]
Knapsack looks really useful but when reading the docs I don't see the type
enforcement, which is the biggest reason I use this Collection library.
Would I create a typed collection with this library then pass it into the
Knapsack collection?

**Comment by icanhazstring** [comment]
Also knapsack doesn't seem to be much taken care of.
Most changes date over two year back.

**Comment by icanhazstring** [comment]
> Does it make sense to require knapsack and use its CollectionTrait?
Could be. But  there could be a problem since knapsack does allow 5.6 and above where we allow 7.2 and above.

I think if there were to be an language/package update on knapsack this might fit better.
And I can't say that I am familiar with knapsack. I just took a quick peek into the implementation.

**Comment by ramsey** [comment]
The PHP version would only be a problem if it were the other way around. It won’t be a problem with knapsack supporting earlier versions of PHP.

**Comment by icanhazstring** [comment]
> The PHP version would only be a problem if it were the other way around. It won’t be a problem with knapsack supporting earlier versions of PHP.

Good point. I will check it out if it is possible without rewiring too much.
Will keep you updated.

**Comment by icanhazstring** [comment]
@ramsey I digged into the `ContainerTrait` as you mentioned. Turns out, that the trait and the `Container` implementation is very hardwired. It uses plain functions coming from an include file. But the functions itself is doing mostly the same as we suggested.

Manipulating a Collection inside Knapsack will return a new `Collection` from a functions. So I see no possibility to "only" use the trait. We would need to use the Collection from Knapsack, extract the items again and use our `Collection` class. But doing this will ruin every possible performance gain which was proposed by @shadowhand.

So the "only" advantage I see is to take the example from Knapsack into account when developing this repository.

**Comment by shadowhand** [comment]
@icanhazstring right, in our custom Collection class (which we are hoping to replace), we use Knapsack functions to manipulate the collection, and then create a new Collection from the result. The benefit is that Knapsack has a wide range of carefully developed functions that are highly efficient and don't need to reinvented.

**Comment by icanhazstring** [comment]
> we use Knapsack functions to manipulate the collection, and then create a new Collection

The problem with that. You would completely lose every advantage that using Knapsack would bring. A small code example:

```
// Typed collection with Awesome::class
$collection = new Ramsey\Collection(Awesome::class);
// Should return everything that does not match the filter function
$filterCollection = $collection->filter(function() { ... });
```

Then this should use the knapsack::filter function.
But the functions are hard wired to knapsack.

```
class AbstractCollection
{
    use Knapsack\CollectionTrait;

    public function filter(callable $callable): Ramsey\CollectionInterface
    {
        // call knapsack filter function, which will return a knapsack\Collection
        $filtered = filter($this->toArray, $callable);
        return new static(iterator_to_array($filtered->getIterator());
    }
}
```

So what you are suggesting is that `Ramsey\Collection` should extend from `Knapsack\Collection` if I am not wrong. This would mean:

1. A complete rewrite of this library - since both libraries have different ideas and different interfaces (typed vs non-typed)
2. Make this library deprecated and add typed collection support to knapsack
3. Copy the functions from knapsack and make them "working" with this library instead

---

I see your point here to not reinvent the wheel. But you can't drive a car with with three default 17" rims and one high performance formular 1 rim which does not even fit onto your car.

**Comment by shadowhand** [comment]
The correct option for performance would be:

```php
return new static(Knapsack\toArray($filtered));
```

I am not suggesting that `Ramsey\Collection` extend `Knapsack\Collection`, that wouldn't make sense and frankly doesn't work, because Knapsack functions always return the same type.

Filter is a trivial case that can be done with native `array_filter`. There are other, more complex cases, that Knapsack already covers. `partitionBy()` would be a reasonable example.

**Comment by shadowhand** [comment]
I don't think the goal should be to implement every Knapsack method. I think `map`, `filter`, `first`, `last` are easy enough, no need to use Knapsack. Really, I think my recommendation is:

Don't implement a ton of methods in this package. Refer users to Knapsack for complex operations (like merge, intersect, etc) and show them how to use it to build new collections.

**Comment by jmauerhan** [comment]
Honestly, I'm really not happy with that as a solution - if I want to run `unique` or `merge` on an existing FooCollection, how would I use Knapsack to do that in a way that's really any different than the problem I laid out in the issue? How does Knapsack let me keep my FooCollection without having a bunch of extra code in my project, that could be part of the TypedCollection?

I'm really not following your suggestion, can you please show me the code you're suggesting we tell other users of this library to use (those users right now being me, since I asked for this functionality). I did an example of sort in the issue, can you show me how to do this with Knapsack in a way that's as succinct as the way I laid out?

I also think it's okay for there to be more than one library that solves the same challenge, in different ways.

**Comment by shadowhand** [comment]
@jmauerhan valid question, and I don't have a good answer. I think for all the use cases you laid out, Knapsack is not a great fit. My intention was simple: don't implement too many methods for complex corner cases when Knapsack exists for complex, multi-step cases. It got pretty side-tracked into "where do we draw the line" and that is 🙅‍♂️ .

tl;dr: the original methods suggested make sense to have in this package.

**Comment by ramsey** [comment]
I’ve toyed with the idea of copying over Knapsack’s function file, leaving their license and copyright intact, and modifying it for this library. Not sure if that’s a good idea, but it would use their work while tailoring it to be used with this library.

**Comment by icanhazstring** [comment]
@ramsey guess this can be closed as well with #44
